File: CPP/testecpp/associacao.cpp
#include <iostream>
#include <string>
using namespace std;

class Caneta {
public:
    Caneta() {}
    Caneta(string name) : nome(name) {}
    void escrever() { cout << "Caneta esta escrevendo" << endl; }
    
    string nome;
};

class MaquinaDeEscrever {
public:
    MaquinaDeEscrever() {}
    void escrever() { cout << "Maquina esta escrevendo" << endl;; }
};

class Escritor {
public:
    Escritor(string name) : nome(name) {}
    void escrever() { ferramenta.escrever(); }
    void metodoAssociativo(Caneta tool) { ferramenta = tool; }
    
    string nome;
    Caneta ferramenta;
};

int main() {
    Escritor escritor {"Joao"};
    Caneta caneta {"Bic"};
    MaquinaDeEscrever maquina;
    escritor.metodoAssociativo(caneta);
    escritor.ferramenta.escrever();
    return 0;
}

File: CPP/testecpp/associacao.py
# coding: utf-8
class Escritor:
    def __init__(self, nome):
        self.nome = nome
        self.ferramenta = None

    def metodo_associativo(self, ferramenta):
        self.ferramenta = ferramenta

class Caneta:
    def __init__(self, nome):
        self.nome = nome
    
    def escrever(self):
        print('Caneta está escrevendo...')

class MaquinaDeEscrever:
    def escrever(self):
        print('Máquina está escrevendo...')      
    
escritor = Escritor('João')
caneta = Caneta('Bic')
maquina = MaquinaDeEscrever()

escritor.metodo_associativo(caneta)
escritor.ferramenta.escrever()
escritor.metodo_associativo(maquina)
escritor.ferramenta.escrever()